File: Emulator/src/main/content/global/skill/summoning/familiar/npc/PackYakNPC.java
package content.global.skill.summoning.familiar.npc;

import content.global.skill.summoning.SummoningScroll;
import content.global.skill.summoning.familiar.BurdenBeast;
import content.global.skill.summoning.familiar.Familiar;
import content.global.skill.summoning.familiar.FamiliarSpecial;
import core.api.Container;
import core.game.node.entity.combat.equipment.WeaponInterface;
import core.game.node.entity.player.Player;
import core.game.node.entity.player.info.LogType;
import core.game.node.entity.player.info.PlayerMonitor;
import core.game.node.item.Item;
import core.game.system.config.ItemConfigParser;
import core.game.world.update.flag.context.Animation;
import core.game.world.update.flag.context.Graphics;
import core.plugin.Initializable;
import org.rs.consts.NPCs;

import static core.api.ContentAPIKt.addItem;
import static core.api.ContentAPIKt.removeItem;

/**
 * The type Pack yak npc.
 */
@Initializable
public class PackYakNPC extends BurdenBeast {

    /**
     * Instantiates a new Pack yak npc.
     */
    public PackYakNPC() {
        this(null, 6873);
    }

    /**
     * Instantiates a new Pack yak npc.
     *
     * @param owner the owner
     * @param id    the id
     */
    public PackYakNPC(Player owner, int id) {
        super(owner, id, 5800, 12093, 12, 30, WeaponInterface.STYLE_AGGRESSIVE);
    }

    @Override
    public Familiar construct(Player owner, int id) {
        return new PackYakNPC(owner, id);
    }

    @Override
    protected boolean specialMove(FamiliarSpecial special) {
        Player player = owner;
        Item item = new Item(special.getItem().getId(), 1);
        if (item.getId() == SummoningScroll.WINTER_STORAGE_SCROLL.getItemId()) {
            return false;
        }
        if (!item.getDefinition().getConfiguration(ItemConfigParser.BANKABLE, true)) {
            player.sendMessage("A magical force prevents you from banking this item.");
            return false;
        }
        Item remove = item;
        if (!item.getDefinition().isUnnoted) {
            remove = new Item(item.getId(), 1);
            item = new Item(item.getNoteChange(), 1);
        }
        boolean success = addItem(player, item.getId(), item.getAmount(), Container.BANK);
        if (success) {
            success = removeItem(player, remove, Container.INVENTORY);
            if (!success) {
                boolean recovered = removeItem(player, item, Container.BANK);
                if (recovered) {
                    PlayerMonitor.log(player, LogType.DUPE_ALERT, "Successfully recovered from potential dupe attempt involving the winter storage scroll");
                } else {
                    PlayerMonitor.log(player, LogType.DUPE_ALERT, "Failed to recover from potentially successful dupe attempt involving the winter storage scroll");
                }
            }
        }
        if (success) {
            player.getDialogueInterpreter().close();
            graphics(Graphics.create(1358));
            player.getPacketDispatch().sendMessage("The pack yak has sent an item to your bank.");
        } else {
            player.getPacketDispatch().sendMessage("The pack yak can't send that item to your bank.");
        }
        return true;
    }

    @Override
    public void visualizeSpecialMove() {
        owner.visualize(Animation.create(7660), Graphics.create(1316));
    }

    @Override
    public int[] getIds() {
        return new int[]{NPCs.PACK_YAK_6873, NPCs.PACK_YAK_6874};
    }

    @Override
    protected String getText() {
        return "Baroo!";
    }

}

File: Server/src/main/content/global/skill/summoning/familiar/npc/PackYakNPC.kt
package content.global.skill.summoning.familiar.npc

import content.global.skill.summoning.SummoningScroll
import content.global.skill.summoning.familiar.BurdenBeast
import content.global.skill.summoning.familiar.Familiar
import content.global.skill.summoning.familiar.FamiliarSpecial
import core.api.Container
import core.api.addItem
import core.api.removeItem
import core.api.sendMessage
import core.game.node.entity.combat.equipment.WeaponInterface
import core.game.node.entity.player.Player
import core.game.node.entity.player.info.LogType
import core.game.node.entity.player.info.PlayerMonitor.log
import core.game.node.item.Item
import core.game.system.config.ItemConfigParser
import core.game.world.update.flag.context.Animation
import core.game.world.update.flag.context.Graphics
import core.plugin.Initializable
import shared.consts.Items
import shared.consts.NPCs

@Initializable
class PackYakNPC @JvmOverloads constructor(owner: Player? = null, id: Int = NPCs.PACK_YAK_6873) :
    BurdenBeast(owner, id, 5800, Items.PACK_YAK_POUCH_12093, 12, 30, WeaponInterface.STYLE_AGGRESSIVE) {
    override fun construct(owner: Player, id: Int): Familiar {
        return PackYakNPC(owner, id)
    }

    override fun specialMove(special: FamiliarSpecial): Boolean {
        val player = owner ?: return false
        var item = special.item ?: return false
        item = Item(item.id, 1)

        if (item.id == SummoningScroll.WINTER_STORAGE_SCROLL.itemId) {
            return false
        }

        if (!item.definition.getConfiguration(ItemConfigParser.BANKABLE, true)) {
            sendMessage(player, "A magical force prevents you from banking this item.")
            return false
        }

        val remove = if (!item.definition.isUnnoted()) {
            Item(item.id, 1)
        } else {
            item
        }

        if (!item.definition.isUnnoted()) {
            item = Item(item.noteChange, 1)
        }

        var success = addItem(player, item.id, item.amount, Container.BANK)
        if (success) {
            success = removeItem(player, remove, Container.INVENTORY)
            if (!success) {
                val recovered = removeItem(player, item, Container.BANK)
                if (recovered) {
                    log(
                        player,
                        LogType.DUPE_ALERT,
                        "Successfully recovered from potential dupe attempt involving the winter storage scroll"
                    )
                } else {
                    log(
                        player,
                        LogType.DUPE_ALERT,
                        "Failed to recover from potentially successful dupe attempt involving the winter storage scroll"
                    )
                }
            }
        }

        if (success) {
            player.dialogueInterpreter.close()
            graphics(Graphics.create(1358))
            sendMessage(player, "The pack yak has sent an item to your bank.")
        } else {
            sendMessage(player, "The pack yak can't send that item to your bank.")
        }
        return true
    }

    override fun visualizeSpecialMove() {
        owner.visualize(Animation.create(7660), Graphics.create(1316))
    }

    override fun getIds(): IntArray {
        return intArrayOf(NPCs.PACK_YAK_6873, NPCs.PACK_YAK_6874)
    }

    override fun getText(): String {
        return "Baroo!"
    }
}
